### PR TITLE
fix(user): 并发建档不再产生重复用户，历史重复行不再阻断消息处理

### LINKED ISSUE #312: [Bug]: 用户表多出来一个用户
### 版本信息

v2.3.3

### 部署方式

Docker 一键部署（含 Napcat）

### 运行环境

Linux

### 系统架构

amd64 (x86_64)

### Python 版本

_No response_

### 发生了什么？

大概就是运行的时候，注册多注了一个用户，导致那个用户一发言就一直报日志bug。

暂时没发现有别的风险。

Image

Image

### 相关信息

```shell

```

### 贡献意向

None

### 补充信息

_No response_

### PR DESCRIPTION
**Pull request by greenhandzdl**
# 🌟 PR 提交说明

## 📋 许可确认

- [x] 我已仔细阅读并同意 [NekroAgent 开源协议](../../LICENSE) 中的所有条款并理解我的贡献将受到该协议的约束，特别是"贡献者条款"部分
- [x] 我确认我提交的代码不会侵犯任何第三方的知识产权

## 💯 代码质量确认

- [x] 我已执行静态类型检查，确保代码不含基本类型错误：
  - 后端：`uv run poe typecheck`（basedpyright，0 errors）+ `uv run poe lint`（ruff）
  - 前端：本次未涉及前端代码
- [x] 无新增类型忽略标记

说明：`poe lint` 仍报 1 处 `tests/test_web_chat_mcp_client.py:1 I001`，这是 main 上既有问题（本 PR 未触碰该文件），未顺带修改以保持变更单一目的。

## 🔍 变更类型

- [x] 🐛 Bug 修复 (修复一个问题)

## 📄 PR 描述

修复 #312「用户表多出来一个用户」：库里出现同一平台用户的重复行，且该用户每次发言都持续刷错误日志。

**根因（已在本机复现，不是推测）**

1. `user` 表在 (`adapter_key`, `platform_userid`) 上**没有唯一约束**（见 `migrations/models/0_20260209003607_init.py` 的建表语句）。
2. `user_register` 是「先 `get_or_none` 查、再 `create` 插」。同一新用户连发两条消息时，NoneBot 把两个事件作为并发任务跑在同一个事件循环上，两者都在对方插入完成前查不到用户 → 各插一行。
3. 重复行一旦出现，`DBUser.get_by_union_id` 用的 `get_or_none` 会抛 `MultipleObjectsReturned: Multiple objects returned for "DBUser", expected exactly one`；而 `collect_message` 中这次读取**不在 try 内**，异常直接冒出消息处理链 → 该用户此后每条消息刷一次错误日志，即 #312 描述的现象。
4. 再调用 `user_register` 也不会自愈：它内部的 `get_or_none` 同样抛 `MultipleObjectsReturned`，而不是设计上的「用户已存在」。

复现证据（真实 Tortoise + SQLite，两条并发注册走同一段生产代码）：

```text
gather done in 0.44s -> ['NoneType', 'NoneType']
ROW COUNT for one platform user = 2  ids=[1, 2]
get_by_union_id RAISED MultipleObjectsReturned: Multiple objects returned for "DBUser", expected exactly one
second user_register RAISED MultipleObjectsReturned: ...
```

**修复**

- `user_register`：查与插在同一事件循环内串行化，从源头不再产生重复行。锁按 `asyncio.get_running_loop()` 取并存在 `WeakKeyDictionary` 里——模块级 `asyncio.Lock` 会绑死在首个循环上，换循环后复用直接 `RuntimeError`（这点是测试跑出来的，不是预防性设计）。
- `DBUser.get_by_union_id`：确定性返回最早一行（`order_by("id").first()`），让**已经带着重复行的现存部署立刻恢复**，不必等数据清理。
- `collect_message`：并发建档的落败方遇到 `ConflictError` 时复用已建好的用户继续处理消息（不再 `logger.exception` 并丢弃消息），并去掉原先的 `assert user`。
- 同类入口一并收口：`adapters/onebot_v11/matchers/message.py`、`adapters/onebot_v11/matchers/notice.py`、`adapters/minecraft/matchers/notice.py` 三处原本直接 `DBUser.get_or_none(adapter_key=..., platform_userid=...)`，同样会撞 `MultipleObjectsReturned`，统一改走 `get_by_union_id`。

**为什么这个 PR 里没有加数据库唯一约束**：唯一索引才是最彻底的一层，但它需要一条「先去重、再 `CREATE UNIQUE INDEX`」的迁移，而迁移必须用 `poe db-revision` 对着真实数据库生成（仓库规则禁止手写迁移文件）。我这边没有可用的 PostgreSQL 凭据（本地 PG 端口通但配置里是占位账号），生成不了也验证不了，所以没把未验证的迁移塞进 PR。如果维护者认为该补，我可以在你们的环境上补做：去重保留 `min(id)`，且 `user` 表目前没有任何外键引用，删除多余行不会造成悬挂引用。

## 🔗 相关 Issue

Closes #312

## 📸 修改效果截图

无 UI 变更。

## 🛠️ 实现复杂度

- [x] 中等 (需要一些技术考量)

## 📋 实现细节

三层各司其职：读取容错（救存量部署）→ 写入串行（不再新增重复）→ 管线复用（并发落败方不丢消息）。不改表结构、不改任何对外语义：`user_register` 对「已存在」仍抛 `ConflictError`，`routers/user.py` 与 `routers/user_manager.py` 的行为不变。

## 🧪 测试步骤

1. 新增用例：`uv run pytest tests/test_user_register_idempotent.py -v`，6 条全部写真库（Tortoise + SQLite），不 mock DB。
   - `test_concurrent_register_creates_single_user`、`test_concurrent_messages_from_new_user_register_once`：并发建档只留一行，且两条消息都不被丢掉；前者还断言「恰好一个成功 + 一个抛 `ConflictError`」，把 collector 赖而复用行为的异常类型钉住；
   - `test_union_lookup_returns_earliest_row_when_duplicates_exist`、`test_collect_message_survives_duplicate_user_rows`：存量重复行下发言不再抛错；
   - `test_register_reports_conflict_when_duplicates_exist`：重复行下再注册报「已存在」而非 DB 异常；
   - `test_rejected_registration_keeps_a_log_line`：建档被拒绝（如昵称撞上保留名 `admin`）时消息丢弃必须留日志。
   用例在改动前全部失败：前五条分别失败于「2 行用户」与 `MultipleObjectsReturned`，最后一条失败于「消息被丢弃但没有任何日志」（它针对的是本 PR 初版自己引入的静默丢弃，由 6f411e1 修掉）。
2. 回归：`uv run pytest`（242 passed）、`uv run poe typecheck`（0 errors）、`uv run bot --load-test`（exit 0）。
3. 实机端到端需要两个 QQ 号或同一账号连发两条消息，本环境无 Napcat 链路，未做实机验证；但并发建档与重复行发言这两条路径已由真实 DB + 真实 `collect_message` 覆盖。

## 🔁 评审收敛记录

- **a192822**（响应 review）：并发建档用例原本丢弃 `asyncio.gather` 的异常，只核对最终行内容。已双向验证这条断言确有分量——把落败方改抛 `OperationFailedError`（即「靠唯一索引兜底」那种看着也正确的实现）后，改前的用例**依然通过**，改后的用例精确报 `assert 0 == 1`。
- **6f411e1**（自查，非 review 要求）：初版 `except ConflictError → 复用查到的行` 把「保留用户名被拒」也当成并发落败，导致昵称叫 `admin` 的用户每条消息被无声丢弃，比改动前连一条日志都没有更难查。现在能复用就复用，不能复用记 warning。
- **b3ffbb7**（响应 review）：注册锁仅保证单进程内互斥。核实了本项目不存在该意见声明的触发条件（`run_bot.py` 无 `workers`、compose 无 `replicas` 且写死 `container_name`、代码内无 `multiprocessing`/第二事件循环），因此不改造成分布式锁，而是把这个前提写进 `services/user/util.py` 的注释；跨实例的兜底仍指向上面点名的唯一约束局限。

## Sourcery 总结

防止平台用户重复，并确保在包含历史重复记录的部署中，消息处理保持健壮。

错误修复：
- 防止并发用户注册创建重复用户，并在注册发生竞争失败时允许消息处理继续进行。
- 使用户查找能够容忍历史重复记录，并通过确定性地选择最早的记录来处理这些重复项。
- 通过支持重复项安全处理的联合 ID 查找来执行适配器用户查找。
- 在注册被拒绝且消息必须丢弃时，保留警告日志。

测试：
- 添加基于数据库的测试，覆盖并发注册、并发消息传递、历史重复记录、冲突处理以及被拒绝的注册。

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Prevent duplicate platform users and keep message handling resilient for deployments containing historical duplicate rows.

Bug Fixes:
- Prevent concurrent user registration from creating duplicate users and allow message processing to continue when registration loses a race.
- Make user lookups tolerate historical duplicate rows by deterministically selecting the earliest record.
- Route adapter user lookups through the duplicate-safe union-ID lookup.
- Preserve warning logs when registration is rejected and the message must be discarded.

Tests:
- Add database-backed tests covering concurrent registration, concurrent message delivery, historical duplicates, conflict handling, and rejected registrations.

</details>

### CODE REVIEW

**Comment by greenhandzdl** [comment]
@sourcery-ai review

（补充上下文：本次修复的是 issue #312 的根因链——`user` 表在 (adapter_key, platform_userid) 上没有唯一约束，`user_register` 先查后插存在 TOCTOU，并发发言会插入重复行；重复行存在时 `DBUser.get_by_union_id` 的 `get_or_none` 抛 `MultipleObjectsReturned`，导致该用户每条消息都处理失败。修复分三层：登录查询改为固定取最早行、建档串行加锁（锁按事件循环存放）、collector 捕获 ConflictError 后复用已存在的行。回归用例在 `tests/test_user_register_idempotent.py`。）

**Comment by greenhandzdl** [comment]
@sourcery-ai review

File: tests/test_user_register_idempotent.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    )
+    second = await db.create(
+        username="第二次建档", password="", adapter_key="onebot_v11", platform_userid=platform_userid, perm_level=0, login_time=now
+    )
+    return first, second
+
+
+async def test_concurrent_register_creates_single_user(user_db: Any) -> None:
+    """同一平台用户的并发建档只允许留下一行"""
+    await asyncio.gather(_register("10001"), _register("10001"), return_exceptions=True)
```
[the remark this answers is not in the document]
Comment (greenhandzdl) [reply tests/test_user_register_idempotent.py]: 这条不是误报，采纳了，谢谢——而且它比我原以为的更有分量。

我按“把实现改错，看用例是否还绿”验证了一遍：把 `user_register` 里落败方抛的 `ConflictError` 换成 `OperationFailedError`（也就是“靠数据库唯一索引兜底、把 IntegrityError 包装成通用失败”那种看着也能防重复的实现），改动前的用例**依然通过**（只剩一行、用户名也对），改动后精确地红在：

```
assert sum(isinstance(result, ConflictError) for result in results) == 1
E   assert 0 == 1
```

之所以这条断言对本 PR 是必要的而不只是风格问题：`collect_message` 侧只认 `ConflictError` 才会去复用已存在的行（`nekro_agent/adapters/interface/collector.py`），落败方抛别的异常就等于把那条消息直接丢掉。所以“恰好一个成功 + 一个 ConflictError”正是这个修复对外承诺的契约，原用例没把它钉住。已按建议改法提交在 a192822。

另外顺带一提，同一个异常契约在另一侧还有个坑我也一并修了（6f411e1）：`user_register` 还会因为保留用户名 `admin` 抛 `ConflictError`，而我原先写的 `except ConflictError → 复用查到的行` 会把这种情况当成“并发落败”，`get_by_union_id` 返回 `None` 后就静默丢消息，比改动前连一条 `logger.exception` 都没有更难查。现在区分成“查得到就复用，查不到记 warning”。

**Comment by greenhandzdl** [comment]
@sourcery-ai review

已处理上一轮的唯一一条意见（并发建档落败方的异常类型断言，见 a192822），另外自行补了一个同源缺陷的修复（建档被拒绝时不再静默丢消息，见 6f411e1）。请重新审查。

**Comment by greenhandzdl** [comment]
@sourcery-ai review

File: nekro_agent/services/user/util.py [hunk 2]
```diff
@@ -20,25 +22,41 @@
...
-            platform_userid=data.platform_userid,
-            perm_level=Role.User,
-            login_time=datetime.now(),
-        )
-    except Exception as e:
-        logger.error(f"注册用户时发生错误: {e}")
-        raise OperationFailedError(operation="注册用户") from e
+    # user 表上没有 (adapter_key, platform_userid) 唯一约束，先查后插必须整体串行；
+    # 否则同一用户并发发言时两个任务都查不到而各插一行，之后该用户每条消息都会读失败。
+    async with _user_register_lock():
```
[the remark this answers is not in the document]
Comment (greenhandzdl) [reply nekro_agent/services/user/util.py:51]: 这条我核实过部署形态后**部分采纳**：它指出的“进程内锁跨不住多进程”在事实层面成立，但给的触发条件（“应用运行于多个工作进程”）在本项目现有的启动与部署路径里不存在，因此不是 issue #312 的复现路径；而它建议的真正兜底手段（唯一约束）正是我在 PR 描述里已点名的局限。

**核实过程（均为读代码/配置得证，不是推测）：**

1. 生产入口 `run_bot.py` 走 NoneBot 原生 `nonebot.run()`，同进程承载 FastAPI；Python 侧 `workers|gunicorn` 零命中，唯一的 `uvicorn.run(` 在 run_bot.py:141，属 `--reload` 开发路径且未传 `workers`（单进程）。
2. `nekro_agent`、`plugins` 内没有 `multiprocessing` / `os.fork` / `Process(target=...)`，也没有 `asyncio.run(` / `new_event_loop` / `run_until_complete`——同一进程内不存在第二个跑 `user_register` 的事件循环。大量 `asyncio.to_thread` / `run_in_executor` 只把**同步函数**丢进线程，协程仍回主循环，构造不出它说的“多事件循环并发注册”。
3. `docker/docker-compose.yml` 里 `nekro_agent` 只有一个实例、无 `replicas`/`scale`，且写死 `container_name: ${INSTANCE_NAME}nekro_agent`——同名容器本身无法横向扩。
4. issue #312 的报发现场也是单容器（Docker + Napcat），报错来自同一进程内两条并发消息“先查后插”的竞态，正是本 PR 修掉的那条路径。

**因此做了什么、没做什么：**
- 做了：把这个前提写进代码注释（services/user/util.py:25-30），明确“仅单进程内互斥；若将来多实例共库，串行化必须落到 user 表 (adapter_key, platform_userid) 的唯一约束上”，避免后来人误把这把锁当跨实例保证。见 b3ffbb7。
- 没做：加唯一约束/迁移。即 PR 描述里“为什么这个 PR 里没有加数据库唯一约束”那段——本项目迁移必须用 `poe db-revision` 对真实 PostgreSQL 生成，我手上没有可用 PG 凭据，未验证过的迁移不会塞进这个 PR。
- 没做：改用 `pg_advisory_lock` 等数据库锁。项目同时支持 sqlite（见 nekro_agent/core/tortoise_config.py 与 db_migration.py 的 postgres/sqlite 分支），为一个当前不存在的触发条件引入方言分支不值当。

若你认为“多实例共库”属于受支持场景，那唯一约束就是必需的，我可以按 `poe db-revision` 补一个“去重 + 加约束”的迁移进来。

**Comment by greenhandzdl** [comment]
@sourcery-ai review

本轮已处理上一条意见：核实了它声明的触发条件在本项目不成立（无 workers/无多实例部署/无第二事件循环），同时把“进程内锁仅单进程互斥”这一前提写进代码注释（b3ffbb7），跨进程兜底仍依赖唯一约束这一点保留在 PR 描述的已知局限里。请重新审查。

**Comment by greenhandzdl** [comment]
@sourcery-ai review

File: nekro_agent/services/user/util.py [hunk 3]
```diff
@@ -20,25 +22,45 @@
...
```
[the remark this answers is not in the document]
Comment (greenhandzdl) [reply nekro_agent/services/user/util.py:51]: 临界区那半句我接受，已改；「按联合用户 ID 分片加锁」这半句我核实后不做，理由和数字如下。

**已改（9af398e）**：`get_hashed_password` 确实原先算在临界区里，与「先查后插」无关却参与串行化，已提到锁外（services/user/util.py）。现在锁只包住 SELECT + INSERT。

**不做的部分，先看实测**（本机 `uv run python`，SQLite 内存库，只计时不改逻辑）：

```
bcrypt('')               = 212.70 ms
SELECT+INSERT under lock =   0.20 ms  per registration
```

也就是说被这条意见点名的「锁把建档串行化」，代价是 0.2 ms 量级；而同一次调用里的 bcrypt 是 212.70 ms，差三个数量级。更关键的是 `get_hashed_password`（services/user/auth.py:17）是**同步函数、直接在协程里调用**，它会把整个事件循环钉住 213 ms——这跟有没有我这把锁无关，把锁拆成 per-user 也一点都收不回来。（PG 上 SELECT+INSERT 会比 0.2 ms 大，我这里没有可用 PG 凭据所以不给未实测的数字；但 213 ms 这一项是不随数据库变的常数。）

再看触发频次：`user_register` 只在一个平台用户**有史以来第一条消息**上执行，稳态消息管线根本不进这把锁。所以分片锁优化的是「大量新用户同时首次发言」这个尖峰，而那个尖峰的真实瓶颈是 bcrypt 阻塞循环，不是锁。

**顺带报一个真问题（本 PR 不动，请维护者定夺）**：上面这条意味着，100 个新用户同时首次发言 ≈ 100 × 213 ms 的事件循环阻塞，全 bot 的消息处理都会被拖住。这是 main 上既有的实现（`user_register` 与 `user_login` 都这么调），不是本 PR 引入的，修法是 `await asyncio.to_thread(get_hashed_password, ...)` 一类。它与 #312 的症状无关，我按「保持单一目的」没塞进来；如果你们认可，我另开 PR 处理，或者你们要我在这个 PR 里一并做也行。

File: tests/test_user_register_idempotent.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+from nekro_agent.services.command.base import CommandPermission
+from nekro_agent.services.user.util import user_register
+
+_USER_ONLY_APPS = {"models": ["nekro_agent.models.db_user"]}
+
+
+@pytest.fixture
+async def user_db():
+    await Tortoise.init(db_url="sqlite://:memory:", modules=_USER_ONLY_APPS)
+    await Tortoise.generate_schemas()
```
[the remark this answers is not in the document]
Comment (greenhandzdl) [reply tests/test_user_register_idempotent.py]: 这条成立，已按仓库规则改（660f97b）。

规则我核对了两处：`.cursor/rules/backend-rules.mdc:69`「禁止调用 `generate_schemas()` 与 `nekro_db_reset`，避免破坏已部署数据」，以及 `.cursor/rules/global.mdc:11`「废弃自动 `generate_schemas`」。并且全仓 `generate_schemas` 的唯一调用方就是我这个新文件，没有任何既有测试给它开先例，所以没有理由保留。

**为什么不是「改走 Aerich 迁移链」**：`migrations/models/0_20260209003607_init.py:222-235` 的 `"user"` 建表语句是 PostgreSQL 方言（`SERIAL` / `TIMESTAMPTZ` / `JSONB`），整个链在 SQLite 上跑不起来；而 `run_db_migrations()` 走的是 `TORTOISE_ORM` 指向的真实库，测试里拿它建表就等于去动部署库。我也没有可用的 PG 凭据（配置里是占位账号）。

**为什么不是「没有 PG 就 skip」**：`.github/workflows/commit.yml` 里 CI 只跑 `uv sync` + `bot --load-test`，压根不跑 pytest。加个 `skipif` 的结果是这 6 条用例在任何地方都不会执行，等于没测。

**改成什么**：按部署那份建表语句手写等价的 SQLite DDL（源已在注释里点名到 migration 文件行号）。这条意见的第二层担心是「验证的是模型自动生成的表而不是部署实际使用的 schema」，所以我补了两组变异验证，证明换 schema 来源没有让用例变弱：

- 把 `DBUser.get_by_union_id` 退回修复前的 `get_or_none`：**3 failed** —— `MultipleObjectsReturned` 这个原始症状照样被抓到；
- 从 DDL 里删掉 `perm_level` 列：**5 failed** —— 模型与建表语句漂移会大声报错，不会静默放过（这点也写进了注释）。

还原后 6 passed；全量 `uv run pytest` 242 passed，`poe typecheck` 0 errors。

**Comment by greenhandzdl** [comment]
@sourcery-ai review

本轮处理了 Copilot 的两条意见：`generate_schemas()` 换成按部署建表语句手写的 DDL（660f97b，并用两组变异验证用例强度未降），密码摘要移出注册锁临界区（9af398e）。分片锁那条按实测数字（bcrypt 212.70ms vs 临界区 0.20ms）说明后未采纳，理由在线程里。请重新审查。
